File: Homework/Hang_Nguyen/lesson_1/single-number.py
def singleNumber(nums):
    result = 0
    for i in range(0, len(nums)):
        result ^= nums[i]
    return result


# TEST CASE
nums = [2,2,1]
res = singleNumber(nums)
print("TEST 1 is " + str(res == 1))

nums = [4,1,2,1,2]
res = singleNumber(nums)
print("TEST 2 is " + str(res == 4))

nums = [1]
res = singleNumber(nums)
print("TEST 3 is " + str(res == 1))

File: Homework/Le_Trong/src/lesson_01/single_number.rs
/**
 * Dùng XOR: Số giống nhau khi XOR sẽ
 * bằng 0. Có thể áp dụng cho 1 dãy số
 * mà các số giống nhau nhưng không liền
 * kề nhau.
 *
 * Ví dụ:
 *  1 ^ 1 = 0
 *  1 ^ 2 ^ 1 ^ 2 = 0
 *
 *  Runtime Complexity: O(n) với n là nums.len()
 *  Space Complexity: O(1)
 */
pub fn single_number(nums: Vec<i32>) -> i32 {
    let mut result = nums[0];
    for i in 1..(nums.len()) {
        result = result ^ nums[i];
    }

    return result;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_number() {
        assert_eq!(single_number(vec![2, 2, 1]), 1);
        assert_eq!(single_number(vec![4, 1, 2, 1, 2]), 4);
        assert_eq!(single_number(vec![1]), 1);
    }
}
